realm/realm-library/src/main/java/io/realm/RealmMap.java

```diff
@@ -0,0 +1,667 @@
+
+import javax.annotation.Nullable;
+
+import io.realm.internal.Freezable;
+import io.realm.internal.ManageableObject;
+import io.realm.internal.OsMap;
+
+/**
+ * RealmMap is used to map keys to values. A RealmMap cannot contain duplicate keys and each key can
+ * map to at most one value. A RealmMap cannot have {@code null} keys but can have {@code null}
+ * values.
+ * <p>
+ * Similarly to {@link RealmList}s, a RealmDictionary can operate in managed and unmanaged modes. In
+ * managed mode a RealmDictionary persists all its contents inside a Realm whereas in unmanaged mode
+ * it functions like a {@link HashMap}.
+ * <p>
+ * Managed RealmDictionaries can only be created by Realm and will automatically update its content
+ * whenever the underlying Realm is updated. Managed RealmDictionaries can only be accessed using
+ * the getter that points to a RealmDictionary field of a {@link RealmObject}.
+ * <p>
+ * Unmanaged RealmDictionaries can be created by the user and can contain both managed and unmanaged
+ * RealmObjects. This is useful when dealing with JSON deserializers like GSON or other frameworks
+ * that inject values into a class. Unmanaged elements in this list can be added to a Realm using
+ * the {@link Realm#copyToRealm(Iterable, ImportFlag...)} method.
+ *
+ * @param <K> the type of the keys stored in this map
+ * @param <V> the type of the values stored in this map
+ */
+public abstract class RealmMap<K, V> implements Map<K, V>, ManageableObject, Freezable<RealmMap<K, V>> {
+
+    protected final MapStrategy<K, V> mapStrategy;
+
+    // ------------------------------------------
+    // Unmanaged constructors
+    // ------------------------------------------
+
+    /**
+     * Instantiates a RealmMap in unmanaged mode.
+     */
+    protected RealmMap() {
+        this.mapStrategy = new UnmanagedMapStrategy<>();
+    }
+
+    /**
+     * Instantiates a RealmMap in unmanaged mode with an initial map.
+     *
+     * @param map initial map.
+     */
+    RealmMap(Map<K, V> map) {
+        this();
+
+        mapStrategy.putAll(map);
+    }
+
+    // ------------------------------------------
+    // Managed constructors
+    // ------------------------------------------
+
+    RealmMap(MapStrategy<K, V> mapStrategy) {
+        this.mapStrategy = mapStrategy;
+    }
+
+    // ------------------------------------------
+    // ManageableObject API
+    // ------------------------------------------
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isManaged() {
+        return mapStrategy.isManaged();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isValid() {
+        return mapStrategy.isValid();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isFrozen() {
+        return mapStrategy.isFrozen();
+    }
+
+    // ------------------------------------------
+    // Map API
+    // ------------------------------------------
+
+    @Override
```

Comment:
Is the Javadoc automatically inherited for all of these methods?

Reply:
It looks like it is.
<img width="828" alt="Screenshot 2021-04-28 at 12 08 13" src="https://user-images.githubusercontent.com/1874445/116386873-7fb33880-a81a-11eb-87b7-e64122ff2cef.png">

However, I experienced that when calling certain methods from the Set interface from Kotlin tests, the documentation shown when hovering over the method call would show the docs from `Collection` and not `Set`. I saw this happening only for Set methods that are routed through Kotlin extension functions, like `public operator fun contains(element: @UnsafeVariance E): Boolean` but then I wrote the `{@inheritDoc}` word and it got solved. All very odd 🤷